### PR TITLE
Implement support for full 24-bit readout mode operations

### PR DESCRIPTION
**Pull request by timcnicholls**
This PR provides support for the full 24-bit readout mode firmware update (version 0x11C). This also incorporates improvements to FEM connection tracking and timeout handling, dual 12-bit readout modes and power card sensor conversion factor fixes.

### CODE REVIEW

**Comment by timcnicholls** [comment]
Closes #10

File: control/fem_api_extension/api/src/ExcaliburFemClient.cpp [hunk 1]
```diff
@@ -872,102 +935,135 @@ void ExcaliburFemClient::stopAcquisition(void)
...
+    FEMLOG(mFemId, logWARNING) << "DMA transfer of " << framesRead
+        << " frames failed to complete in expected time during async stop";
+    FEMLOG(mFemId, logWARNING) << "Expected " << framesRead * subFramesPerFrame 
+      << " BDs TXed, got " << acqState.totalSent 
+      << " RX: top " << acqState.totalRecvTop << " bottom " << acqState.totalRecvBot;
   }
 
+//    }
+//  }
+
```
Comment (GDYendell) [inline control/fem_api_extension/api/src/ExcaliburFemClient.cpp:1053]: What is this?

File: control/fem_api_extension/api/src/ExcaliburFemClient.cpp [hunk 2]
```diff
@@ -979,10 +1075,14 @@ void ExcaliburFemClient::stopAcquisition(void)
...
+    // Clear the stop command bit from the ASIC command word register
+    FEMLOG(mFemId, logDEBUG) << "Clearing stop command from ASIC control";
+    this->asicControlCommandExecute((asicControlCommand)0);
+
     // Reset ASIC control firmware block
-    //FEMLOG(mFemId, logDEBUG) << "Sending ASIC control reset to FEM";
+    FEMLOG(mFemId, logDEBUG) << "Sending reset command to ASIC control";
     this->asicControlReset();
-
+
```
Comment (GDYendell) [inline control/fem_api_extension/api/src/ExcaliburFemClient.cpp:1085]: Spaces

**Review by GDYendell — APPROVED** [review]

File: control/excalibur/client/client.py [hunk 3]
```diff
@@ -58,8 +67,12 @@ class ExcaliburDefinitions(object):
     FEM_LFSR_BYPASS_MODE_ENABLED = 1 
     FEM_LFSR_BYPASS_MODE_NAMES = ('disabled', 'enabled')
     
-    FEM_COUNTER_DEPTH_MAP = {1: 0, 6: 1, 12: 2, 24: 3}
-    
+    FEM_COUNTER_DEPTH_MAP = OrderedDict(sorted(
+        {'1': 0, '6': 1, '12': 2, '24': 3, 'dual12': 4}.items(), 
+        key=lambda t: t[1]
+    ))
```
[the remark this answers is not in the document]
Comment (timcnicholls) [reply control/excalibur/client/client.py:73]: Didn't know you could initialise an OrderedDict like that, it's much simpler. Thanks!

File: control/fem_api_extension/api/src/ExcaliburFemClient.cpp (hunk 1 above)
Comment (timcnicholls) [reply control/fem_api_extension/api/src/ExcaliburFemClient.cpp:1053]: Lonely stray brackets looking for friends... Now removed.

**Review by ajgdls — APPROVED** [review]

**Comment by GDYendell** [comment]
I think the this branch does not have the frameSimulator parameter name changes (to add underscores)

**Comment by timcnicholls** [comment]
> I think the this branch does not have the frameSimulator parameter name changes (to add underscores)

You're right, but isn't that change in `odin-data` later than the release EXCALIBUR is built against? @ajgdls ?? If not, I can easily incorporate it as it's stashed.

**Comment by ajgdls** [comment]
I've just checked the 24bit dev branch I was working on and it is built against odin-data 1-4-0dls4.  This does appear to have those changes (commit "move to underscore" by William Nichols 24/03/2020).

**Comment by timcnicholls** [comment]
@ajgdls Thanks.

I've now pushed that change also.

**Comment by ajgdls** [comment]
Thanks @timcnicholls - sorry I missed that.  I guess in my dev branch I probably merged in the relevant update to make it compile.

File: control/excalibur/client/client.py [hunk 4]
```diff
@@ -67,11 +67,9 @@ class ExcaliburDefinitions(object):
...
     FEM_LFSR_BYPASS_MODE_NAMES = ('disabled', 'enabled')
     
-    FEM_COUNTER_DEPTH_MAP = OrderedDict(sorted(
-        {'1': 0, '6': 1, '12': 2, '24': 3, 'dual12': 4}.items(), 
-        key=lambda t: t[1]
-    ))
-    FEM_COUNTER_DEPTH_NAMES = FEM_COUNTER_DEPTH_MAP.keys()
+    FEM_COUNTER_DEPTH_MAP = OrderedDict(
+        (('1', 0), ('6', 1), ('12', 2), ('24', 3), ('dual12', 4))
+    )   FEM_COUNTER_DEPTH_NAMES = FEM_COUNTER_DEPTH_MAP.keys()
```
Comment (GDYendell) [inline control/excalibur/client/client.py:72]: This is not valid. I guess this isn't checked in the Travis build?

Comment (GDYendell) [reply control/excalibur/client/client.py:72]:
```suggestion
    FEM_COUNTER_DEPTH_MAP = OrderedDict(
        (('1', 0), ('6', 1), ('12', 2), ('24', 3), ('dual12', 4))
    )
    FEM_COUNTER_DEPTH_NAMES = FEM_COUNTER_DEPTH_MAP.keys()
```

Comment (timcnicholls) [reply control/excalibur/client/client.py:72]: It's trying to use the suggested change feature that messed this up. One minute I'll push the correct one.

**Pull request merged**